### DISCUSSION TITLE
[Question] Must MQTTAsync_destroy be called after the disconnect operation completed?

### DISCUSSION THREAD

**Discussion by yeajay0001** [eclipse-paho/paho.mqtt.c#1444] [Q&A]
Hi,

In paho.mqtt.c example, MQTTAsync_destroy is called after disconnect operation completed: https://github.com/eclipse/paho.mqtt.c/blob/master/src/samples/MQTTAsync_subscribe.c

```
while (!disc_finished)
 	{
		#if defined(_WIN32)
			Sleep(100);
		#else
			usleep(10000L);
		#endif
 	}

destroy_exit:
	MQTTAsync_destroy(&client);
```

And paho.mqtt.cpp example is also called after disconnect operation completed：https://github.com/eclipse/paho.mqtt.cpp/blob/master/src/samples/async_consume_v5.cpp
```
cli.stop_consuming();
cli.disconnect()->wait();
```

And I have two questions:
1、Must MQTTAsync_destroy be called after the disconnect operation completed?
2、Must MQTTAsync_destroy be called in the same thread of call MQTTAsync_create?

Best Regards.

**Comment by icraggs** [comment]
1. You don't need to call destroy if you're happy for the program to end and let the OS clean up resources.
2. No it doesn't have to be called on the same thread as the create was, but you do have to be sure that all the background processing has stopped (by letting disconnect complete fully for instance).
